### PR TITLE
Feat/file domain

### PR DESCRIPTION
**Pull request by fredericvilcot**
👉 This PR adds a new domain into the current hexagonal architecture (`File`) that implements https://github.com/okp4/ui/issues/393.

💡 The idea here is to store files (and especially their `stream` part) to upload them later on a cloud file provider like `ipfs` or `minio`.

🧐 This PR embeds a polyfill for [web-streams](https://github.com/MattiasBuelens/web-streams-polyfill) because `jest` uses `jsdom` as DOM environment that does not implement `stream` lib natively...

### CODE REVIEW

**Review by amimart — CHANGES_REQUESTED** [review]
Seems good so far, but I have a critical remark though..

File: src/domain/file/usecase/clear-file/actionCreators.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+/* eslint-disable @typescript-eslint/explicit-function-return-type */
+import type { ActionsUnion } from 'domain/common/store.helper'
+import { createAction } from 'domain/common/store.helper'
+
+export const ClearFilections = {
```
[the remark this answers is not in the document]
Comment (fredericvilcot) [reply src/domain/file/usecase/clear-file/actionCreators.ts:5]: 👀
If you don't mind, I will fix the typo and I propose to upgrade the naming of files (to singular) in a second step through a refactoring PR

**Review by lolottetheclash — APPROVED** [review]
That's a great Job 👍

**Review by JoeyDjr — APPROVED** [review]

**Review by amimart — APPROVED** [review]

**Review by ccamel — COMMENTED** [review]
Last minutes remarks. Saw that you had merged, but the following remarks should be taken into consideration as they are structural. Don't hesitate if you have any questions.

**Comment by fredericvilcot** [comment]
@ccamel 
Thanks for you remarks, they all make sense.
I will open a refactor pull request ASAP to take into account those changes.

**Comment by ccamel** [comment]
@fredericvilcot Forgot to tell you that of course your PR is a great improvement and the tests are so sexy. 😄

File: src/domain/file/usecase/save-files/saveFiles.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  const errorToDispatch = ErrorMapper.mapRawErrorToEntity(error)
+  dispatch(ThrowErrorActions.errorThrown(errorToDispatch))
+}
+
+export const saveFiles =
+  (files: DeepReadonly<SaveFilesPayload<string>>): ThunkResult<Promise<void>> =>
+  // eslint-disable-next-line @typescript-eslint/typedef
+  async (dispatch, getState) => {
+    files.forEach((file: DeepReadonly<SaveFilePayload<string>>) => {
+      if (getState().file.byId.has(file.id)) {
```
[the remark this answers is not in the document]
Comment (fredericvilcot) [reply src/domain/file/usecase/save-files/saveFiles.ts:20]: Sounds good, I need some extra work to better understand the command pattern execution in a DDD context

File: src/domain/file/usecase/clear-files/clearFiles.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import type { ThunkResult } from 'domain/file/store/store'
+import { ClearFilesActions } from './actionCreators'
+
+export const clearFiles =
+  (): ThunkResult<Promise<void>> =>
+  // eslint-disable-next-line @typescript-eslint/typedef
+  async dispatch => {
+    dispatch(ClearFilesActions.filesCleared())
```
[the remark this answers is not in the document]
Comment (fredericvilcot) [reply src/domain/file/usecase/clear-files/clearFiles.ts:8]: @ccamel 
I fully agree with you, the exemple you wrote down is cristal clear and highlights the predominance of the generated events and the way to handle commands.
If I understand the changes, as we made for `storeFiles` use-case, the `removeFiles` use-case should dispatch a `fileRemoved` event as many times as there are files in the store?

[the remark this answers is not in the document]
Comment (ccamel) [reply src/domain/file/usecase/clear-files/clearFiles.ts:8]: > @ccamel 
> If I understand the changes, as we made for `storeFiles` use-case, the `removeAllFiles` use-case should dispatch a `fileRemoved` event as many times as there are files in the store?
> 

Yes! That's it.

**Pull request merged**